### PR TITLE
support repos.yaml default repo 'ref'

### PR DESCRIPTION
**Pull request by jhoblitt**

### CODE REVIEW

File: python/lsst/ci/prepare.py [hunk 1]
```diff
@@ -162,12 +162,48 @@ def _origin_candidates(self, product):
...
+        if self.repos and product in self.repos:
+
+            spec = self.repos[product]
+            if isinstance(spec, str):
+                origin = spec
+            elif isinstance(spec, dict):
+                origin = spec['url']
+                ref = spec['ref']
+            else:
+                raise Exception('lp0 on fire')
```
Comment (ktlim) [inline python/lsst/ci/prepare.py]: Cute, but maybe something a little more descriptive?  (Yes, I know the auto-linter checks this, but maybe it still got corrupted locally somehow...)

File: python/lsst/ci/prepare.py [hunk 2]
```diff
@@ -162,12 +162,48 @@ def _origin_candidates(self, product):
...
+        (origin, ref) = self._repos_yaml_coordinates(product)
+
+        if origin:
+            locations.append(origin)
         if self.repository_patterns:
             locations += [ pat % data for pat in self.repository_patterns ]
         return locations
 
+    def _ref_candidates(self, product):
+        """ Generate a list of refs to to attempt to checkout. """
```
Comment (ktlim) [inline python/lsst/ci/prepare.py]: Too many "to"s.

File: python/lsst/ci/prepare.py [hunk 3]
```diff
@@ -162,12 +162,48 @@ def _origin_candidates(self, product):
         """ Expand repository_patterns into URLs. """
         data = { 'product': product }
         locations = []
-        if self.repos and product in self.repos:
-            locations.append(self.repos[product])
+        (origin, ref) = self._repos_yaml_coordinates(product)
```
Comment (timj) [inline python/lsst/ci/prepare.py]: parens aren't needed

File: python/lsst/ci/prepare.py [hunk 4]
```diff
@@ -162,12 +162,48 @@ def _origin_candidates(self, product):
...
             locations += [ pat % data for pat in self.repository_patterns ]
         return locations
 
+    def _ref_candidates(self, product):
+        """ Generate a list of refs to to attempt to checkout. """
+
+        # ref precedence should be:
+        # user specified refs > repos.yaml default ref > implicit master
+        refs = self.refs
+        (origin, ref) = self._repos_yaml_coordinates(product)
```
Comment (timj) [inline python/lsst/ci/prepare.py]: Parens

File: python/lsst/ci/prepare.py (hunk 3 above)
Comment (jhoblitt) [inline python/lsst/ci/prepare.py]: You can take the developer of perl5 but...

**Pull request merged**